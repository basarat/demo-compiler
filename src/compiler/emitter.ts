import { BinaryExpressionNode, CallExpressionNode, NumericLiteralNode, Program, Node } from './types';

export function emitter(program: Program): string {
  function emit(node: Node): string {
    if (node.type === 'NumericLiteral') {
      return emitNumericLiteral(node);
    } else if (node.type === 'BinaryExpression') {
      return emitBinaryExpression(node);
    } else if (node.type === 'CallExpression') {
      return emitCallExpression(node);
    } else {
      throw new SyntaxError('Unknown Node');
    }
  }

  function emitNumericLiteral(node: NumericLiteralNode) {
    return node.value;
  }

  function emitBinaryExpression(node: BinaryExpressionNode) {
    return `${emit(node.left)} ${node.operator.type === 'PlusToken' ? '+' : '-'} ${emit(node.right)}`;
  }

  function emitCallExpression(node: CallExpressionNode) {
    if (node.identifier.value === 'log') {
      return `console.log(${emit(node.argument)})`;
    } else {
      throw new SyntaxError(`Unknown Identifier in call expression: ${node.identifier}`);
    }
  }

  const output: string[] = [];
  for (const node of program.body) {
    output.push(emit(node));
  }
  return output.join('\n');
}