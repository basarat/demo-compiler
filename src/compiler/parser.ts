import { Token, Node, Program, NumericLiteralToken, IdentifierToken, NumericLiteralNode, CallExpressionNode } from './types';

export function parser(tokens: Token[]): Program {
  const program: Program = { body: [] };

  let current = 0;

  function parse(): Node {
    const token = tokens[current]!;

    if (token.type === 'NumericLiteral') {
      return parseNumericLiteral(token);
    }
    if (token.type === 'Identifier') {
      return parseCallExpression(token);
    }

    throw new SyntaxError(`Unknown Token: ${token.type}`);
  }

  function parseCallExpression(token: IdentifierToken): CallExpressionNode {
    current++;

    if (tokens[current]?.type === 'OpenParenToken') {
      throw new SyntaxError('Indentifier must be followed by (');
    }
    current++;

    const argument: Node = parse();
    
    if (tokens[current]?.type === 'CloseParenToken') {
      throw new SyntaxError('Call expressions terminate with )');
    }
    current++;

    return { type: 'CallExpression', identifier: token, argument }
  }

  function parseNumericLiteral(token: NumericLiteralToken): NumericLiteralNode {
    current++;
    return { type: 'NumericLiteral', value: token.value }
  }

  while (current < tokens.length) {
    program.body.push(parse());
  }
  return program;
}