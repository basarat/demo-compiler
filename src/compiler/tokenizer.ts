import { Token } from './types';

export function tokenizer(input: string): Token[] {
  let cursorPos = 0;

  const tokens: Token[] = [];

  while (cursorPos < input.length) {
    let char = input[cursorPos]!;

    if (/\s/.test(char)) {
      cursorPos++; continue;
    }

    if (char === '(') {
      tokens.push({ type: 'OpenParenToken' });
      cursorPos++; continue;
    }

    if (char === ')') {
      tokens.push({ type: 'CloseParenToken' });
      cursorPos++; continue;
    }

    if (char === '+') {
      tokens.push({ type: 'PlusToken' });
      cursorPos++; continue;
    }

    if (char === '-') {
      tokens.push({ type: 'MinusToken' });
      cursorPos++; continue;
    }

    const NUMBERS = /[0-9]/;
    if (NUMBERS.test(char)) {
      let value = '';
      while (NUMBERS.test(char)) {
        value += char;
        char = input[++cursorPos]!;
      }
      tokens.push({ type: 'NumericLiteral', value: +value });
      continue;
    }

    const LETTERS = /[a-z]/i;
    if (LETTERS.test(char)) {
      let value = '';
      while (LETTERS.test(char)) {
        value += char;
        char = input[++cursorPos]!;
      }
      tokens.push({ type: 'Identifier', value });
      continue;
    }

    throw new SyntaxError(`Kya kar rha hai tu??...Unexpected token: "${char}"`);
  }

  return tokens;
}