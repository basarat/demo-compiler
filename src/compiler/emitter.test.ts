import { emitter } from './emitter';
import { parser } from './parser';
import { tokenizer } from './tokenizer';
test('it should pass', () => {
  expect(emitter(parser(tokenizer('log(1+3-4)')))).toMatchSnapshot();
});
