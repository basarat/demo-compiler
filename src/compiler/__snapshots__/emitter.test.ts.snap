// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should pass 1`] = `"console.log(1 + 3 - 4)"`;
